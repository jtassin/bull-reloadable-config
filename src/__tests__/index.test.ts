import { reloadConfig, BullConfig } from ".."
import Queue from 'bull';

describe('bull-reloadable-config', () => {

    const VERSION_0_0_4 = '0.0.4';
    let queueName = 'testQueueName';;
    const queueOptions = {};
    const extraQueueName = 'extraQueueName';
    let queue = new Queue(queueName);
    const extraQueue = new Queue(extraQueueName);

beforeEach(async () => {
    await queue.close()
    queueName = Math.random() * 10000 + 'random'
    queue = new Queue(queueName);
})

    afterEach(async () => {
        await queue.clean(0)
        await queue.empty();
        await extraQueue.empty();
    })

    afterAll(async () => {
        await queue.close()
        await extraQueue.close()
    })

    it('throws if a job as no options', async () => {
        const configs: BullConfig[] = [
             //@ts-ignore
            {data: { _version: VERSION_0_0_4 },        }]
        await expect(reloadConfig(queueName, queueOptions, configs)).rejects.toMatchSnapshot()
    })
    it('throws if a job as no _id', async () => {
        const configs: BullConfig[] = [
             //@ts-ignore
            {opts: {}, data: { _version: VERSION_0_0_4 },        }]
        await expect(reloadConfig(queueName, queueOptions, configs)).rejects.toMatchSnapshot()
    })

    it('throws if two jobs has the same ids', async () => {
        const configs: BullConfig[] = [{
            data: { _version: VERSION_0_0_4 },
            opts: { jobId: 'id'},
        }, {
            data: {  _version: VERSION_0_0_4 },
            opts: { jobId: 'id'},
        }]
        await expect(reloadConfig(queueName, queueOptions, configs)).rejects.toMatchSnapshot()
    })

    it('creates a non existing job', async () => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        const job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: VERSION_0_0_4, ...extraData })
    })

    it('transmits correctly the job options', async () => {
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: VERSION_0_0_4 },
            opts: { jobId, attempts: 5, delay: 8},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        const job = await queue.getJob(jobId);
        expect(job!.opts).toMatchObject({ attempts: 5, delay: 8 })
    })

    it('transmits correctly the job name', async () => {
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            name: 'Heisenberg',
            data: {  _version: VERSION_0_0_4 },
            opts: { jobId, attempts: 5, delay: 8},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        const job = await queue.getJob(jobId);
        expect(job!.name).toBe('Heisenberg')
    })

    it('recreate a job in the higher version with changes', async() => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        let job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: VERSION_0_0_4, ...extraData })

        await reloadConfig(queueName, queueOptions, [{
            data: {  _version: '0.0.5', ...extraData },
            opts: { jobId, attempts: 5, delay: 8},
        }
        ])
        job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: '0.0.5', ...extraData })
        expect(job!.opts).toMatchObject({ attempts: 5, delay: 8 })
    })

    it('does not recreate a job in the lower version', async() => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: '0.0.5', ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        let job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: '0.0.5', ...extraData })

        await reloadConfig(queueName, queueOptions, [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ])
        job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: '0.0.5', ...extraData })
    })

    it('can recreate a job in the lower version if force is true', async() => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: '0.0.5', ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        let job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: '0.0.5', ...extraData })

        await reloadConfig(queueName, queueOptions, [{
            force: true,
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ])
        job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: VERSION_0_0_4, ...extraData })
    })

    it('remove jobs not in the config only if deleteExtraJobs is true', async () => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const otherJobId = 'otherJobId'
        const configs: BullConfig[] = [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        let job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: VERSION_0_0_4, ...extraData })

        await reloadConfig(queueName, queueOptions, [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId: otherJobId},
        }
        ], false)
        job = await queue.getJob(jobId);
        expect(job).toBeTruthy()

        await reloadConfig(queueName, queueOptions, [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId: otherJobId},
        }
        ], true)
        job = await queue.getJob(jobId);
        expect(job).toBeFalsy()
    })

    it('can create huge job array in a reasonable time', async () => {
        jest.setTimeout(15_000);
        const count = 10_000;
        let configs: BullConfig[] = [
        ];
        for(let i = 0 ; i < count / 2 ; i+=1) {
            configs.push({
                data: {  _version: VERSION_0_0_4 },
                opts: { jobId: 'i' + i },
            })
        }
        await reloadConfig(queueName, queueOptions, configs)
        configs = [
            {
                data: {  _version: VERSION_0_0_4 },
                opts: { jobId: 'd1', delay: 3600},
            },
            {
                data: {  _version: VERSION_0_0_4 },
                opts: { jobId: 'd2', delay: 8},
            }
        ];
        for(let i = 0 ; i < count ; i+=1) {
            configs.push({
                data: {  _version: '0.0.5' },
                opts: { jobId: 'i' + i },
            })
        }
        await reloadConfig(queueName, queueOptions, configs)

        const jobCount = await queue.getJobCounts();
        expect(jobCount).toEqual({
            "active": 0,
            "completed": 0,
            "delayed": 2,
            "failed": 0,
            "paused": 0,
            "waiting": count,
        })
    })

    it('refuses non semver version', async () => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: 'thisisbadversion', ...extraData },
            opts: { jobId},
        }
        ];
        await expect(reloadConfig(queueName, queueOptions, configs)).rejects.toMatchSnapshot()

    })

    it('does not remove and add a job that did not change', async () => {
        const extraData = { a: 'g'}
        const jobId = 'myjobid'
        const configs: BullConfig[] = [{
            data: {  _version: VERSION_0_0_4, ...extraData },
            opts: { jobId},
        }
        ];
        await reloadConfig(queueName, queueOptions, configs)
        let job = await queue.getJob(jobId);
        const previousTimestamp = job!.timestamp;
        expect(job!.data).toEqual({ _version: VERSION_0_0_4, ...extraData })

        await reloadConfig(queueName, queueOptions, [{
            data: {  _version: '0.0.5', ...extraData },
            opts: { jobId},
        }
        ])
        job = await queue.getJob(jobId);
        expect(job!.data).toEqual({ _version: '0.0.5', ...extraData })
        expect(job!.timestamp).toBe(previousTimestamp)
    })

})