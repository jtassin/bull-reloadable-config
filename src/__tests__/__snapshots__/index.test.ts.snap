// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bull-reloadable-config refuses non semver version 1`] = `[SyntaxError: thisisbadversion is not a valid semver]`;

exports[`bull-reloadable-config throws if a job as no _id 1`] = `[SyntaxError: Each job need an _id]`;

exports[`bull-reloadable-config throws if a job as no options 1`] = `[SyntaxError: Each job need an _id]`;

exports[`bull-reloadable-config throws if two jobs has the same ids 1`] = `[SyntaxError: Each job need an uniq _id found id twice]`;
